lib/spack/spack/binary_distribution.py

```diff
@@ -1388,42 +1386,21 @@ def extract_tarball(spec, filename, allow_root=False, unsigned=False,
     buildinfo = spec_dict.get('buildinfo', {})
     old_relative_prefix = buildinfo.get('relative_prefix', new_relative_prefix)
     rel = buildinfo.get('relative_rpaths')
-    # if the original relative prefix and new relative prefix differ the
-    # directory layout has changed and the  buildcache cannot be installed
-    # if it was created with relative rpaths
     info = 'old relative prefix %s\nnew relative prefix %s\nrelative rpaths %s'
     tty.debug(info %
               (old_relative_prefix, new_relative_prefix, rel))
-#    if (old_relative_prefix != new_relative_prefix and (rel)):
-#        shutil.rmtree(tmpdir)
-#        msg = "Package tarball was created from an install "
-#        msg += "prefix with a different directory layout. "
-#        msg += "It cannot be relocated because it "
-#        msg += "uses relative rpaths."
-#        raise NewLayoutException(msg)
-
-    # extract the tarball in a temp directory
+
+    # Extract the tarball into the store root, presumably on the same filesystem.
+    # The directory created is the base directory name of the old prefix.
+    # Moving the old prefix name to the new prefix location should preserve
+    # hard links and symbolic links.
+    extracted_dir = os.path.join(spack.store.layout.root,
+                                 old_relative_prefix.split(os.path.sep)[-1])
```

Comment:
Can we do the temporary extraction somewhere else? I'm a bit worried about having leftovers of a failed decompression that need to be removed manually from the store.

Reply:
I can put a try except around it.

Reply:
But it needs to be the same filesystem as the store root so shutil.move handles hard links correctly.

Reply:
@gartung Can we at least extract these in a `.tmp` or `.buildcache` subdirectory of the store before moving them? That would also allow to easily clean that place with `spack clean`.

Reply:
Just saw this. I will add the subdirectory as suggested.

Reply:
Thanks!